hooks/pre/31-upgrade-puppet.rb

```diff
@@ -47,6 +47,7 @@ def fail_and_exit(message)
 end
 
 if app_value(:upgrade_puppet)
+  fail_and_exit 'Puppet already installed and upgraded. Skipping.' if File.exist?('/etc/foreman-installer/.puppet_4_upgrade')
```

Comment:
I'd prefer if the file location would be a constant somewhere. That removes the risk of mismatches.